### PR TITLE
More user-friendly RSS URL generation with extra parameters

### PR DESCRIPTION
**Pull request by TomAbel**

### CODE REVIEW

File: upload/admin/controller/module/smaily_for_opencart.php [hunk 1]
```diff
@@ -141,7 +147,32 @@ public function index() {
...
+        }
+        if ($this->config->get('smaily_for_opencart_rss_sort_by')) {
+            $data['smaily_rss_url'] .= '&sort_by=' . $this->config->get('smaily_for_opencart_rss_sort_by');
+        }
+        if ($this->config->get('smaily_for_opencart_rss_sort_order')) {
+            $data['smaily_rss_url'] .= '&sort_order=' . $this->config->get('smaily_for_opencart_rss_sort_order');
+        }
+        if ($this->config->get('smaily_for_opencart_rss_limit')) {
+            $data['smaily_rss_url'] .= '&limit=' . $this->config->get('smaily_for_opencart_rss_limit');
+        }
```
Comment (kaittodesk) [inline upload/admin/controller/module/smaily_for_opencart.php:163]: URL-encode values passed to the URL.

File: upload/admin/view/template/module/smaily_for_opencart.tpl [hunk 2]
```diff
@@ -321,6 +319,84 @@
                 </div>
               </div>
             </div>
+            <!-- RSS -->
+            <div id="section4" class="tab-pane fade in">
+              <div class="form-group">
+                <label class="col-sm-2 control-label" for="rss-category"><?php echo $rss_category_title; ?></label>
+                <div class="col-sm-10">
+                  <select name="smaily_for_opencart_rss_category" id="rss-category" class="form-control smaily-rss-options">
+                    <option value=''>All Products</option>
```
Comment (kaittodesk) [inline upload/admin/view/template/module/smaily_for_opencart.tpl:328]: Formatting: mixing single and double quotes in HTML.

File: upload/admin/view/template/module/smaily_for_opencart.tpl [hunk 3]
```diff
@@ -321,6 +319,84 @@
...
+                    <?php
+                    foreach ($rss_categories as $category) {
+                      $selected = $category['category_id'] === $rss_category ? 'selected' : '';
+                      echo("<option value=$category[category_id] $selected>$category[name]</option>");
+                    }
```
Comment (kaittodesk) [inline upload/admin/view/template/module/smaily_for_opencart.tpl:333]: Use pattern described here https://github.com/sendsmaily/smaily-opencart-module/pull/94#discussion_r482091656 to rework this code block.

File: upload/admin/view/template/module/smaily_for_opencart.tpl [hunk 4]
```diff
@@ -321,6 +319,84 @@
...
+                  <select name="smaily_for_opencart_rss_sort_by" id="rss-sort-by" class="form-control smaily-rss-options">
+                   <?php
+                    // All available options
+                    $sort_options = [
+                      'pd.name' => $sort_name,
+                      'p.model' => $sort_model,
+                      'p.price' => $sort_price,
+                      'p.status' => $sort_status,
+                      'p.sort_order' => $sort_order
+                    ];
```
Comment (kaittodesk) [inline upload/admin/view/template/module/smaily_for_opencart.tpl:350]: Provide this mapping in the controller.

File: upload/admin/view/template/module/smaily_for_opencart.tpl [hunk 5]
```diff
@@ -321,6 +319,84 @@
...
+                    // Add options for select.
+                    foreach ($sort_options as $sort_code => $sort_name) {
+                      $selected = $rss_sort_by === $sort_code ? 'selected' : '';
+                      echo("<option value='$sort_code' $selected>$sort_name</option>");
+                    }
```
Comment (kaittodesk) [inline upload/admin/view/template/module/smaily_for_opencart.tpl:355]: Use the same pattern here, as with category field.

File: upload/admin/view/template/module/smaily_for_opencart.tpl [hunk 6]
```diff
@@ -496,6 +572,35 @@
...
+      if (rss_limit != "") {
+        parameters.limit = rss_limit;
+      }
+      $('#smaily-rss-feed-url').html(rss_url_base + $.param(parameters));
+    });
   });
 })(jQuery);
 </script>
+<script type="text/javascript">
+    var smaily_rss_url_base = '<?php echo $smaily_rss_url_base; ?>';
```
Comment (kaittodesk) [inline upload/admin/view/template/module/smaily_for_opencart.tpl:605]: Are you using the `smaily_rss_url_base` variable outside of the jQuery closure? If not, then you can move the variable definition before `change` handler:

```javascript
var smaily_rss_url_base = '<?php echo $smaily_rss_url_base; ?>';
$(".smaily-rss-options").change(function (event) {
    // ...
});
```

**Review by kaittodesk — CHANGES_REQUESTED** [review]

File: upload/admin/controller/module/smaily_for_opencart.php (hunk 1 above)
Comment (TomAbel) [reply upload/admin/controller/module/smaily_for_opencart.php]: Using `http_build_query`

File: upload/admin/view/template/module/smaily_for_opencart.tpl (hunk 2 above)
Comment (TomAbel) [reply upload/admin/view/template/module/smaily_for_opencart.tpl:326]: Fixed this. Let me know if I missed any

File: upload/admin/view/template/module/smaily_for_opencart.tpl (hunk 3 above)
Comment (TomAbel) [reply upload/admin/view/template/module/smaily_for_opencart.tpl]: Not echoing out HTML anymore

File: upload/admin/view/template/module/smaily_for_opencart.tpl (hunk 4 above)
Comment (TomAbel) [reply upload/admin/view/template/module/smaily_for_opencart.tpl]: Added

File: upload/admin/view/template/module/smaily_for_opencart.tpl (hunk 5 above)
Comment (TomAbel) [reply upload/admin/view/template/module/smaily_for_opencart.tpl]: Not echoing out HTML

File: upload/admin/view/template/module/smaily_for_opencart.tpl (hunk 6 above)
Comment (TomAbel) [reply upload/admin/view/template/module/smaily_for_opencart.tpl]: Nope, moved it there like in OC 1.5

File: upload/admin/controller/module/smaily_for_opencart.php [hunk 7]
```diff
@@ -141,7 +147,37 @@ public function index() {
...
+            $query_parameters['sort_by'] = $this->config->get('smaily_for_opencart_rss_sort_by');
+        }
+        if ($this->config->get('smaily_for_opencart_rss_sort_order')) {
+            $query_parameters['sort_order'] = $this->config->get('smaily_for_opencart_rss_sort_order');
+        }
+        if ($this->config->get('smaily_for_opencart_rss_limit')) {
+            $query_parameters['limit'] = $this->config->get('smaily_for_opencart_rss_limit');
+        }
+        $data['smaily_rss_url'] = $this->config->get('config_url') . 'index.php?route=smailyforopencart/rss&'
+            . http_build_query($query_parameters);
```
Comment (kaittodesk) [inline upload/admin/controller/module/smaily_for_opencart.php:165]: You could D.R.Y-up RSS URL compilation:

```suggestion
        $data['smaily_rss_url'] = $data['smaily_rss_url_base'] . '&' . http_build_query($query_parameters);
```

File: upload/admin/view/template/module/smaily_for_opencart.tpl [hunk 8]
```diff
@@ -275,10 +272,9 @@
                     ];
                     // Add options for select.
                     foreach ($cart_options as $value => $name) {
-                      $selected = is_array($abandoned_additional) && in_array($value, $abandoned_additional) ? 'selected' : '';
-                      echo("<option value='$value' $selected>$name</option>");
-                    }
-                    ?>
+                      $selected = is_array($abandoned_additional) && in_array($value, $abandoned_additional) ? 'selected' : ''; ?>
+                      <option value="<?php echo $value; ?>"<?php echo $selected; ?>><?php echo $name; ?></option>
```
Comment (kaittodesk) [inline upload/admin/view/template/module/smaily_for_opencart.tpl:276]: Missing whitespace between `value` and `selected` attributes.

File: upload/admin/view/template/module/smaily_for_opencart.tpl [hunk 9]
```diff
@@ -321,6 +317,74 @@
...
+            <div id="section4" class="tab-pane fade in">
+              <div class="form-group">
+                <label class="col-sm-2 control-label" for="rss-category"><?php echo $rss_category_title; ?></label>
+                <div class="col-sm-10">
+                  <select name="smaily_for_opencart_rss_category" id="rss-category" class="form-control smaily-rss-options">
+                    <option value="">All Products</option>
+                    <?php
+                    foreach ($rss_categories as $category) {
+                      $selected = $category['category_id'] === $rss_category ? 'selected' : ''; ?>
+                      <option value="<?php echo $category['category_id']; ?>"<?php echo $selected; ?>><?php echo $category['name']; ?></option>
```
Comment (kaittodesk) [inline upload/admin/view/template/module/smaily_for_opencart.tpl:330]: Missing whitespace between `value` and `selected` attributes.

File: upload/admin/view/template/module/smaily_for_opencart.tpl [hunk 10]
```diff
@@ -321,6 +317,74 @@
...
+              </div>
+              <div class="form-group">
+                <label class="col-sm-2 control-label" for="rss-sort-by"><?php echo $rss_sort_by_title; ?></label>
+                <div class="col-sm-10">
+                  <select name="smaily_for_opencart_rss_sort_by" id="rss-sort-by" class="form-control smaily-rss-options">
+                   <?php
+                    // Add options for select.
+                    foreach ($sort_options as $sort_code => $sort_name) {
+                      $selected = $rss_sort_by === $sort_code ? 'selected' : ''; ?>
+                      <option value="<?php echo $sort_code; ?>"<?php echo $selected; ?>><?php echo $sort_name; ?></option>
```
Comment (kaittodesk) [inline upload/admin/view/template/module/smaily_for_opencart.tpl:343]: Missing whitespace between `value` and `selected` attributes.

File: upload/admin/view/template/module/smaily_for_opencart.tpl [hunk 11]
```diff
@@ -321,6 +317,74 @@
...
+                  </div>
+                  <?php if ($error_limit) { ?>
+                      <div class="text-danger"><?php echo $error_limit; ?></div>
+                  <?php } ?>
+                </div>
+              </div>
+              <div class="form-group">
+                <label class="col-sm-2 control-label"><?php echo $rss_feed_title; ?></label>
+                <div class="col-sm-10">
+                  <p><strong id="smaily-rss-feed-url" name="rss_feed_url"><?php echo $smaily_rss_url; ?></strong></p>
```
Comment (kaittodesk) [inline upload/admin/view/template/module/smaily_for_opencart.tpl:383]: HTML's `<strong>` tag does not have `name` attribute. What is its purpose?

**Review by kaittodesk — CHANGES_REQUESTED** [review]

File: upload/admin/controller/module/smaily_for_opencart.php (hunk 7 above)
Comment (TomAbel) [reply upload/admin/controller/module/smaily_for_opencart.php]: Yes, good idea, added

File: upload/admin/view/template/module/smaily_for_opencart.tpl [hunk 12]
```diff
@@ -199,10 +197,9 @@
                     ];
                     // Add options for select.
                     foreach ($sync_options as $value => $name) {
-                      $selected = is_array($syncronize_additional) && in_array($value, $syncronize_additional) ? 'selected' : '';
-                      echo("<option value='$value' $selected>$name</option>");
-                    }
-                    ?>
+                      $selected = is_array($syncronize_additional) && in_array($value, $syncronize_additional) ? 'selected' : ''; ?>
+                      <option value="<?php echo $value; ?>"<?php echo $selected; ?>><?php echo $name; ?></option>
```
[the remark this answers is not in the document]
Comment (TomAbel) [reply upload/admin/view/template/module/smaily_for_opencart.tpl]: This is fixed

File: upload/admin/view/template/module/smaily_for_opencart.tpl (hunk 8 above)
Comment (TomAbel) [reply upload/admin/view/template/module/smaily_for_opencart.tpl]: This is fixed

File: upload/admin/view/template/module/smaily_for_opencart.tpl (hunk 10 above)
Comment (TomAbel) [reply upload/admin/view/template/module/smaily_for_opencart.tpl]: This is fixed

File: upload/admin/view/template/module/smaily_for_opencart.tpl (hunk 11 above)
Comment (TomAbel) [reply upload/admin/view/template/module/smaily_for_opencart.tpl]: No purpose, removed.

**Review by kaittodesk — CHANGES_REQUESTED** [review]
LGTM, but the same issue as with `1.5.x-1.2.0`- the target branch should be named `dev/2.2.x-1.4.0`.

File: upload/admin/view/template/module/smaily_for_opencart.tpl (hunk 4 above)
Comment (kaittodesk) [reply upload/admin/view/template/module/smaily_for_opencart.tpl]: Yeah, views should contain as little logic as possible - they should be used only to display content and have minimal conditional logic. But let's not make these changes within this PR.

**Review by kaittodesk — APPROVED** [review]

**Pull request merged**